### PR TITLE
Refactor OFFER/ACCEPT uTP stream and add `portal_*Offer` json-rpc endpoint

### PR DESCRIPTION
**Pull request by ogenev**
### What was wrong?
-  Offer/Accept implementation was not tested.
- `portal_*Offer` json-rpc endpoint was not implemented.

### How was it fixed?
- Manually tested Offer/Accept uTP stream in both directions with Fluffy and fixed all known issues.
- Implemented `portal_*Offer` json-rpc endpoint.

[//]: # (Stay ahead of things, add list items here!)
- [x] Add entry to the [release notes](https://github.com/ethereum/trin/blob/master/newsfragments/README.md) (may forgo for trivial changes)
- [x] Clean up commit history

### CODE REVIEW

File: trin-state/src/jsonrpc.rs [hunk 1]
```diff
@@ -81,6 +82,26 @@ impl StateRequestHandler {
                     };
                     let _ = request.resp.send(response);
                 }
+                StateEndpoint::Offer => {
+                    let response = match OfferParams::try_from(request.params) {
+                        Ok(val) => {
+                            let content_keys =
+                                val.content_keys.iter().map(|x| x.to_vec()).collect();
```
Comment (njgheorghita) [inline trin-state/src/jsonrpc.rs:89]: *nitpick* but i'd prefer using something like `key` rather than `x` here and in `trin-history`

**Review by njgheorghita — APPROVED** [review]
Lookin good!

File: trin-core/src/utp/stream.rs [hunk 2]
```diff
@@ -274,20 +274,14 @@ impl UtpListener {
...
-                                        debug!("Store {key:?}, {content:?}");
-                                    }
-                                }
-                                Err(_) => debug!("Recv malformed data on handing UtpAccept"),
-                            }
+                        UtpMessageId::AcceptStream(content_keys) => {
+                            // TODO: Implement this with overlay store and decode receiver stream if multiple content values are send
+                            debug!("Store {content_keys:?}, {received_stream:?}");
+                        }
+                        UtpMessageId::FindContentData(_content) => {
```
Comment (jacobkaufmann) [inline trin-core/src/utp/stream.rs:281]: Wouldn't content data via uTP stream correspond to `UtpMessageId::FindContentStream`? Maybe I misunderstand the difference between the two variants.

File: trin-core/src/portalnet/overlay.rs [hunk 3]
```diff
@@ -398,30 +403,18 @@ impl<TContentKey: OverlayContentKey + Send, TMetric: Metric + Send>
             message: content_items,
         };
 
-        let utp_listener = self.utp_listener.clone();
         tokio::spawn(async move {
-            if let Some(conn) = utp_listener
-                .write()
+            // send the content to the acceptor over a uTP stream
+            if let Err(msg) = conn
```
Comment (jacobkaufmann) [inline trin-core/src/portalnet/overlay.rs:408]: nit: I think `err` is a more clear variable name here and in other `Err` matches.

File: trin-core/src/portalnet/types/messages.rs [hunk 4]
```diff
@@ -491,6 +491,13 @@ pub struct Accept {
     pub content_keys: BitList<typenum::U8>,
 }
 
+impl Into<Value> for Accept {
+    fn into(self) -> Value {
+        // TODO: self.content_keys returns a number of set bits to the json-rpc client, probably we would like to return a list of 0 and 1 bits
+        serde_json::json!({ "connection_id": format!("{:?}", self.connection_id.to_be()) , "content_keys": self.content_keys.num_set_bits()})
```
Comment (jacobkaufmann) [inline trin-core/src/portalnet/types/messages.rs:497]: I think we can go ahead and return the hex-string for `content_keys`, which seems to be the type in the initial JSON-RPC spec.

File: trin-core/src/utp/stream.rs [hunk 5]
```diff
@@ -717,7 +711,7 @@ impl UtpSocket {
             // When SYN is send and we receive STATE, do not reply
             (SocketState::SynSent, PacketType::State) => {
                 self.connected_to = src;
-                self.ack_nr = packet.seq_nr();
+                self.ack_nr = packet.seq_nr() - 1;
```
Comment (jacobkaufmann) [inline trin-core/src/utp/stream.rs:714]: It may be better to address this deviation from the BEP29 spec in a separate, tiny PR as it seems unrelated to the rest. I don't feel too strongly about this though.

File: trin-core/src/portalnet/overlay_service.rs [hunk 6]
```diff
@@ -661,6 +661,9 @@ impl<TContentKey: OverlayContentKey + Send, TMetric: Metric + Send>
             .map_err(|e| OverlayRequestError::AcceptError(e))?;
         let connection_id: u16 = crate::utp::stream::rand();
 
+        // TODO: Pipe this with overlay DB and request only not available keys.
+        let accept_keys = request.content_keys.clone();
```
[the remark this answers is not in the document]
Comment (ogenev) [reply trin-core/src/portalnet/overlay_service.rs:665]: I will create an issue for this because there is a `should_store` dummy method that is used a few lines below which needs to be addressed too.

File: trin-core/src/utp/stream.rs (hunk 2 above)
Comment (ogenev) [reply trin-core/src/utp/stream.rs:281]: This is kinda tricky because when we respond with the `Content` message with `connection-id`,  we want to keep somewhere the actual content data that we want to stream when we receive `SYN` packet from the requester (the node which sent `FindContent` rpc). 

So right now we keep this data in memory as part of the `UtpMessageId` as `FindContentData(Content)` and we store it on the `send_send_id` channel only, we keep `FindContentStream` for `conn_recv_id`: https://github.com/ethereum/trin/blob/7f66b18221c1124a7e386c193feed5a8a132a0fd/trin-core/src/portalnet/overlay_service.rs#L626

File: trin-core/src/utp/stream.rs (hunk 5 above)
Comment (ogenev) [reply trin-core/src/utp/stream.rs:714]: Sending OFFER messages and init uTP stream to Fluffy breaks if we don't address it here :) I'll create an issue to not forget to add uTP tests that catch those deviations.
